datafusion/core/src/physical_optimizer/sort_enforcement2.rs

```diff
@@ -0,0 +1,2872 @@
+        true
+    }
+}
+
+fn ensure_sorting(
+    requirements: PlanWithSortRequirements,
+) -> Result<Option<PlanWithSortRequirements>> {
+    if let Some(sort_exec) = requirements.plan.as_any().downcast_ref::<SortExec>() {
+        // Remove unnecessary SortExec(local/global)
+        if let Some(result) = analyze_immediate_sort_removal(&requirements, sort_exec) {
+            return Ok(Some(result));
+        }
+    } else if let Some(sort_pres_exec) = requirements
+        .plan
+        .as_any()
+        .downcast_ref::<SortPreservingMergeExec>()
+    {
+        // SortPreservingMergeExec + SortExec(local/global) is the same as the global SortExec
+        // Remove unnecessary SortPreservingMergeExec + SortExec(local/global)
+        if let Some(child_sort_exec) =
+            sort_pres_exec.input().as_any().downcast_ref::<SortExec>()
+        {
+            if sort_pres_exec.expr() == child_sort_exec.expr() {
+                if let Some(result) =
+                    analyze_immediate_sort_removal(&requirements, child_sort_exec)
+                {
+                    return Ok(Some(result));
+                }
+            }
+        } else if !requirements.satisfy_single_distribution
+            || sort_pres_exec
+                .input()
+                .output_partitioning()
+                .partition_count()
+                <= 1
+        {
+            if let Some(result) =
+                analyze_immediate_spm_removal(&requirements, sort_pres_exec)
+            {
+                return Ok(Some(result));
+            }
+        }
+    }
+    let plan = &requirements.plan;
+    let parent_required = requirements.required_ordering.as_deref();
+    if ordering_satisfy_requirement(plan.output_ordering(), parent_required, || {
+        plan.equivalence_properties()
+    }) {
+        // Can satisfy the parent requirements, change the adjusted_request_ordering for UnionExec and WindowAggExec(BoundedWindowAggExec)
+        if let Some(union_exec) = plan.as_any().downcast_ref::<UnionExec>() {
+            // UnionExec does not have real sort requirements for its input. Here we change the adjusted_request_ordering to UnionExec's output ordering and
+            // propagate the sort requirements down to correct the unnecessary descendant SortExec under the UnionExec
+            let adjusted = new_sort_requirements(union_exec.output_ordering());
+            return Ok(Some(PlanWithSortRequirements {
+                required_ordering: None,
+                adjusted_request_ordering: vec![
+                    adjusted;
+                    requirements
+                        .adjusted_request_ordering
+                        .len()
+                ],
+                ..requirements
+            }));
+        } else if plan.as_any().downcast_ref::<WindowAggExec>().is_some()
+            || plan
+                .as_any()
+                .downcast_ref::<BoundedWindowAggExec>()
+                .is_some()
+        {
```

Comment:
You can write a util function `is_window` to check for whether executor is `WindowAggExec` or `BoundedWindowAggExec`. There are couple of place this check is done. I think it will be handy.

Reply:
sure